### ISSUE TITLE
Reduce boilerplate when define same function for Dataset and ChannelGroup

### ISSUE THREAD

**Issue by joefowler** [usnistgov/mass#141]
**[Original report](https://bitbucket.org/joe_fowler/mass/issues/141) by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

We have a lot of functions (5-10) following the pattern

In channel.py
```
class Dataset():
...
def myfun(a,b,c,d,e,f):
    """docstring"""
    dostuff()
```

In channel_group.py
```
class TESChannelGroup():
...
def mfyun(a,b,c,d,e,f):
    for ds in data:
        try:
            ds.myfun(a,b,c,d,e,f)
        except:
            data.set_chan_bad(ds.channum,"myfun failed")
```

I'm pretty sure (and Young thought I was right) that we could change this to
ONLY in channel.py
```
class Dataset():
...
@put_it_in_data
def myfun(a,b,c,d,e,f):
    """docstring"""
    dostuff()
```

I find this makes it kind of painful to add new arguments, since you have to do it in two places. Also almost every function that does this looks different in `channel_group`. And it means the docstring is usually only in one place (most often channel).

Does this seem like it would be worth the effort?

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

I don't understand decorators at all, so it's hard to comment.

Do you mean that somehow that decorator would auto-populate the TESGroup with wrapper functions that do the loop and the try/except ?

I think it does seem worth the effort. My questions would be:

* Does the new auto-made function appear in the help text for the TESgroup, too? Or would it be like a stealth method?
* Who would learn enough about decorators to make this work? Do you know how to do it? Maybe I should, so I can get the hang of it!

Note an additional benefit: it would be good to trap `KeyboardInterrupt` exceptions separately and allow them to fail. If the `TESGroup` methods are made automatically, this would be easy to code in one place. just another example of how this change would help us to follow the DRY principle.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Let me do this one. This is a good idea.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

See https://bitbucket.org/snippets/nist_microcal/gegxxK as an example of how we can do this.

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

When I tried to write down how to do this, it became apparent I don't know how to do it. Young understands decorators.

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

That looks good. What you do think about including the docstring of method with some prepended text like "Loop over each good dataset, on errors mark the dataset bad, apply the following: "

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Good idea. I updated the snippet linked above to do this.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Warning: I cannot imagine that this sort of monkey-patching to the `TESGroup` is going to be reflected in any static documentation that you might want to make (I'm referring to #137 here).

Does that disturb you? Can you think of a sensible workaround?

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

We talked about this, but as far as I understand, the documentation is generated by inspecting objects, not by inspecting code, so it should work as long ass you modify the Class (as your example does).

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

The commit de99df9 seems to work for the automated generation of method `TESGroup.drift_correct`. Nice!  There's one ugly thing about the docstring, however. Look:

```
In [11]: ds=data.channel[59]; g=ds.good()

In [12]: ?data.drift_correct
Signature: data.drift_correct(*args, **kwargs)
Docstring:
Loop over self, calling the drift_correct(...) method for each channel.

drift_correct(...) docstring reads:
Drift correct using the standard entropy-minimizing algorithm
File:      /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/mass-0.6.1-py2.7-macosx-10.12-x86_64.egg/mass/core/channel.py
Type:      instancemethod

In [13]: ?ds.drift_correct
Signature: ds.drift_correct(forceNew=False, category=None)
Docstring: Drift correct using the standard entropy-minimizing algorithm
File:      /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/mass-0.6.1-py2.7-macosx-10.12-x86_64.egg/mass/core/channel.py
Type:      instancemethod
```

Do you notice how the channel has the arguments correctly listed, but the `TESGroup` has only the dummy arguments of `data.drift_correct(*args, **kwargs)`? I'm going to go ahead in spite of this, but this is a slight problem.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Never mind. I think I see the solution in package [`inspect`](https://docs.python.org/2.7/library/inspect.html#classes-and-functions)

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Completed automated TESGroup looping method filter_data.

Fixes #141, though of course more could be done along these lines.
